javalib/src/main/scala/java/net/Inet6Address.scala

```diff
@@ -57,31 +87,62 @@ final class Inet6Address private (
 
   def isIPv4CompatibleAddress(): Boolean = ipAddress.take(12).forall(_ == 0)
 
+  private def formatScopeId(): String = {
+    if (nif != null)
+      nif.getDisplayName()
+    else if (useScopeId) {
+      val netIf = NetworkInterface.getByIndex(scopeId)
+      if (netIf == null) String.valueOf(scopeId)
+      else netIf.getDisplayName()
+    } else ""
+  }
+
 }
 
 object Inet6Address {
 
   def getByAddress(
       host: String,
       addr: Array[Byte],
-      scope_id: Int
+      scopeId: Int
   ): Inet6Address = {
     if (addr == null || addr.length != 16)
       throw new UnknownHostException("Illegal IPv6 address")
 
-    new Inet6Address(addr, host, Math.max(0, scope_id))
+    /* JVM treats negative scopeId as having being not supplied.
+     * Explicitly specified 0 scopeIds are considered supplied.
+     * Elsewhere implicit 0 scopeIds, say from a sin6_scope_id, are not.
+     */
+    if (scopeId < 0)
+      Inet6Address(addr, host)
+    else
+      new Inet6Address(addr, host, true, scopeId, null)
   }
 
-  private[net] def apply(
-      ipAddress: Array[Byte],
+  def getByAddress(
       host: String,
-      scopeId: Int,
-      zone: String
+      addr: Array[Byte],
+      nif: NetworkInterface
+  ): Inet6Address = {
+    if (addr == null || addr.length != 16)
+      throw new UnknownHostException("Illegal IPv6 address")
+
+    /* match JVM
+     * Do not throw on null nif but fail late with obscure/wrong/unexpected
+     * scopeId of 0.
+     */
+
+    new Inet6Address(addr, host, false, 0, nif)
+  }
+
+  private[net] def apply(
+      addr: Array[Byte],
+      host: String
   ): Inet6Address = {
-    new Inet6Address(ipAddress, host, scopeId, zone)
+    new Inet6Address(addr, host, false, 0, null)
   }
 
-  private val hexCharacters = "0123456789abcdef"
+  private final val hexCharacters = "0123456789abcdef"
```

Comment:
Maybe for my understanding, why `final` ? Since the is defined in an object this hasn't any effect right ?

Reply:
Making it final can make it a literal constant in compiler, but for that probably it's name should start with upper case. 
```suggestion
  private final val HexCharacters = "0123456789abcdef"
```

Reply:
RustedBones,

Thank you for looking at this, and in detail.

Yes, my intent in adding `final` was to increase the possibility that a little know & not obvious 
compiler optimization would happen.  I believe Scala itself has no concept of "const".

Wojciech,

Is the intent of making the initial character uppercase so that it can potentially be used in
pattern matching?  That convention for `final val` is new to me, times change and hopefully improve. 
Thank you for kindly picking up on my obsolete ways. 

The suggested change is easy enough to make here. With your consent, I can apply the convention
 i other files as I may touch them. Normally, I would  doing a sweep through the code base looking for
similar usages. I do not feel the benefit justifies the potential breakage, especially in files with which
I am not familiar.  Please advise.